lib/tree.rb

```diff
@@ -16,51 +16,165 @@ def initialize
+  end
+
+  def bst_inorder(node, arr)
+    if node
+      bst_inorder(node.left, arr) if node.left
+      arr << {:key => node.key, :value => node.value}
+      bst_inorder(node.right, arr) if node.right
+    end
+    return arr
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+
+
+  # Time Complexity: O(n) because we visit each node once
+  # Space Complexity: O(h) where h is equal to height of tree unless tree is imbalanced, then O(n)
   def preorder
-    raise NotImplementedError
+    array = []
+    root_node = @root
+    return bst_preorder(root_node, array)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def bst_preorder(node, arr)
+    if !node
+      arr << {:key => node.key, :value => node.value}
+      bst_preorder(node.left, arr) if node.left
+      bst_preorder(node.right, arr) if node.right
+    end
+
+    return arr
+  end
+
+  # Time Complexity: O(n) because we visit each node once
+  # Space Complexity: O(h) where h is equal to height of tree unless tree is imbalanced, then O(n)
   def postorder
-    raise NotImplementedError
+    array = []
+    root_node = @root
+    return bst_postorder(root_node, array)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def bst_postorder(node, arr)
+    if node
+      bst_postorder(node.left, arr) if node.left
+      bst_postorder(node.right, arr) if node.right
+      arr << {:key => node.key, :value => node.value}
+    end
+    return arr 
+  end
+
+  # Time Complexity: O(n) because you have to check each node/subtree in the tree to determine height
+  # Space Complexity: O(logn) unless unbalanced, then O(n)
   def height
-    raise NotImplementedError
+    current = @root
+    return 0 if @root.nil?
+
+    return tree_height(current)
+  
+  end
+
+  def tree_height(node)
+    #find the height of the largest subtree + root node
+    height_left = 0
+    height_right = 0
+
+    if node == nil
+      return 0
+    else
+      height_left = tree_height(node.left) if node.left
+      height_right = tree_height(node.right) if node.right
+      height_left > height_right ? 1 + height_left : 1 + height_right
+    end
   end
 
   # Optional Method
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n) because we visit every node
+  # Space Complexity: O(n) because I am storing all bfs nodes in an array
   def bfs
-    raise NotImplementedError
+    node = @root
+    array = []
+    return breadth_first(node, array)
+  end
+
+  def breadth_first(node, arr)
+    current = node
+    levels_list = []
+    i = 0
+
+    return arr if node.nil?
+    # add level by level
+    while i <= levels_list.length
+      arr << {:key => current.key, :value => current.value} 
+      levels_list << current.left if current.left
+      levels_list << current.right if current.right
+      current = levels_list[i]
+      i+= 1
+    end
+    return arr
   end
 
   # Useful for printing
   def to_s
-    return "#{self.inorder}"
+    puts "#{self.inorder}"
```

Comment:
Just a note, you can't test something being printed, it's easier to test a string result.  

Plus you can always do:  `puts my_tree.to_s`